### PR TITLE
Skeleton task1

### PR DESCRIPTION
**Pull request by yu-krugliak**
Hello, plz review my first task

### CODE REVIEW

File: OnlineGameStore.Api/Controllers/CommentConroller.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+using OnlineGameStore.Application.Services.Interfaces;
+using OnlineGameStore.Infrastructure.Entities;
+
+namespace OnlineGameStore.Api.Controllers
+{
+    [ApiController]
+    [Route("[controller]")]
+    public class CommentConroller : ControllerBase
+    {
+
```
Comment (ataromoku) [inline OnlineGameStore.Api/Controllers/CommentConroller.cs:12]: Remove empty line or add it to all controllers to maintain consistency. First variant is better

File: OnlineGameStore.Api/Controllers/GameController.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        [HttpGet]
+        public async Task<IActionResult> GetAllGames()
+        {
+            return Ok(await _gameService.GetAllAsync());
+        }
+
+        [HttpPost("new")]
+        public async Task<IActionResult> AddGame([FromBody] GameRequest request)
+        {
+            var result = await _gameService.AddAsync(request);
```
Comment (ataromoku) [inline OnlineGameStore.Api/Controllers/GameController.cs:32]: `CreatedAtAction()`

File: OnlineGameStore.Api/Program.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    var connectionString = builder.Configuration.GetConnectionString("DbConnection");
+    options.UseSqlServer(connectionString);
+    
+
+}).AddScoped<GamesContext>();
+
+builder.Services.AddMvc();
+builder.Services.AddMapsterConfiguration()
+        .AddTransient<IMapper, Mapper>();
+builder.Services.AddTransient<IGameRepository, GameRepository>();
```
Comment (ataromoku) [inline OnlineGameStore.Api/Program.cs:35]: Extract adding services to extensions

File: OnlineGameStore.Api/Program.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+// Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
+builder.Services.AddEndpointsApiExplorer();
+builder.Services.AddSwaggerGen();
+
+builder.Services.AddDbContext<GamesContext>((provider, options) =>
+{
+
+    var connectionString = builder.Configuration.GetConnectionString("DbConnection");
+    options.UseSqlServer(connectionString);
+
```
Comment (ataromoku) [inline OnlineGameStore.Api/Program.cs:28]: Why empty lines everywhere!?

File: OnlineGameStore.Application/Mapster/MapsterConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+            return config;
+        }
+
+        public static IServiceCollection AddMapsterConfiguration(this IServiceCollection services)
+        {
+            var config = GetConfiguration();
+            return services.AddSingleton(config);
+        }
+
```
Comment (ataromoku) [inline OnlineGameStore.Application/Mapster/MapsterConfiguration.cs:54]: Empty line

File: OnlineGameStore.Application/Models/Requests/GameRequest.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+using System.Threading.Tasks;
+
+namespace OnlineGameStore.Application.Models.Requests
+{
+    public class GameRequest
+    {
+        public string? Name { get; set; }
+
+        public string? Description { get; set; }
+
```
Comment (ataromoku) [inline OnlineGameStore.Application/Models/Requests/GameRequest.cs:16]: Empty line

File: OnlineGameStore.Application/Models/Views/GameView.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+namespace OnlineGameStore.Application.Models.Views
+{
+    public class GameView
+    {
+        public Guid Key { get; set; }
+
+        public string? Name { get; set; }
+
+        public string? Description { get; set; }
+
```
Comment (ataromoku) [inline OnlineGameStore.Application/Models/Views/GameView.cs:17]: Empty line

File: OnlineGameStore.Application/Services/Implementation/GameService.cs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            var game = _mapper.Map<Game>(gameRequest);
+
+            var addedGame = await _unitOfWork.GameRepository.AddAsync(game);
+
+            return _mapper.Map<GameView>(addedGame);
+        }
+
+        public async Task DeleteByKeyAsync(Guid gameKey)
+        {
+            var result = await _unitOfWork.GameRepository.DeleteByIdAsync(gameKey);
```
Comment (ataromoku) [inline OnlineGameStore.Application/Services/Implementation/GameService.cs:39]: Name variable accordingly (mb isDeleted)

File: OnlineGameStore.Application/Services/Implementation/GameService.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        {
+            var games = await _unitOfWork.GameRepository.GetAllAsync();
+
+            var gamesViews = _mapper.Map<IEnumerable<GameView>>(games);
+            return gamesViews;
+        }
+
+        public async Task<IEnumerable<GameView>> GetByGenre(Guid genreId)
+        {
+            var genre = await _unitOfWork.GenreRepository.GetGenreByIdWithDetails(genreId);
```
Comment (ataromoku) [inline OnlineGameStore.Application/Services/Implementation/GameService.cs:57]: I think we should take games from the game repo, not from genre

File: OnlineGameStore.Application/Services/Implementation/GameService.cs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        {
+            var game = await GetExistingGameById(gameKey);
+
+            var gameView = _mapper.Map<GameView>(game);
+            return gameView;
+        }
+
+        public async Task<IEnumerable<GameView>> GetByPlatform(Guid platformId)
+        {
+            var platformType = await _unitOfWork.PlatformRepository.GetPlatformByIdWithDetails(platformId);
```
Comment (ataromoku) [inline OnlineGameStore.Application/Services/Implementation/GameService.cs:74]: And platforms too

File: OnlineGameStore.Application/Services/Implementation/ServiceBase.cs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+            if (entity is null)
+            {
+                throw new NotFoundException($"{typeof(TEntity).Name} with such id doesn't exist.");
+            }
+
+            return entity;
+        }
+    }
+}
+*/
```
Comment (ataromoku) [inline OnlineGameStore.Application/Services/Implementation/ServiceBase.cs:36]: Uncomment or remove file

File: OnlineGameStore.Application/Services/Interfaces/IGameService.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OnlineGameStore.Application.Services.Interfaces
+{
+    public interface IGameService
+    {
+        Task<IEnumerable<GameView>> GetAllAsync();
+        //Task<IEnumerable<Game>> GetAllWithDetailsAsync();
```
Comment (ataromoku) [inline OnlineGameStore.Application/Services/Interfaces/IGameService.cs:15]: Remove comments

File: OnlineGameStore.Application/Services/Interfaces/IGenreService.cs [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OnlineGameStore.Application.Services.Interfaces
+{
+    public interface IGenreService
+    {
+        Task<GenreView> AddAsync(GenreRequest genreRequest);
+
```
Comment (ataromoku) [inline OnlineGameStore.Application/Services/Interfaces/IGenreService.cs:14]: Empty line

File: OnlineGameStore.Application/Services/Interfaces/IService.cs [hunk 14]
```diff
@@ -0,0 +1,14 @@
+﻿/*using System;
```
Comment (ataromoku) [inline OnlineGameStore.Application/Services/Interfaces/IService.cs:1]: Remove file

File: OnlineGameStore.Application/Services/UnitOfWorkImplementation/UnitOfWork.cs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+        public IGameRepository GameRepository => _gameRepository;
+
+        public ICommentRepository CommentRepository => _commentRepository;
+
+        public IGenreRepository GenreRepository => _genreRepository;
+
+        public IPlatformRepository PlatformRepository => _platformRepository;
+
+        public void Save()
```
Comment (ataromoku) [inline OnlineGameStore.Application/Services/UnitOfWorkImplementation/UnitOfWork.cs:39]: You can use async version of save method

File: OnlineGameStore.Application/Services/UnitOfWorkImplementation/UnitOfWork.cs [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            _gamesContext.SaveChanges();
+        }
+
+        private bool disposed = false;
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!disposed)
```
Comment (ataromoku) [inline OnlineGameStore.Application/Services/UnitOfWorkImplementation/UnitOfWork.cs:48]: Make it less complicated, we don't need two checks

File: OnlineGameStore.Infrastructure/Context/Configurations/CommentConfiguration.cs [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+{
+    public class CommentConfiguration : IEntityTypeConfiguration<Comment>
+    {
+        public void Configure(EntityTypeBuilder<Comment> builder)
+        {
+            builder.HasKey(comment => comment.Id);
+
+            builder.Property(comment => comment.Name);
+            builder.Property(comment => comment.Body);
+            //builder.Property(comment => comment.DatePosted);
```
Comment (ataromoku) [inline OnlineGameStore.Infrastructure/Context/Configurations/CommentConfiguration.cs:15]: Remove comment

File: OnlineGameStore.Infrastructure/Context/Configurations/GameConfiguration.cs [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+namespace OnlineGameStore.Infrastructure.Context.Configurations
+{
+    public class GameConfiguration : IEntityTypeConfiguration<Game>
+    {
+        public void Configure(EntityTypeBuilder<Game> builder)
+        {
+            builder.HasKey(game => game.Id);
+
+            builder.Property(game => game.Key);
+            builder.Property(game => game.Name);
```
Comment (ataromoku) [inline OnlineGameStore.Infrastructure/Context/Configurations/GameConfiguration.cs:14]: Add constraints to text columns

File: OnlineGameStore.Infrastructure/Context/GamesContext.cs [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+    public class GamesContext : DbContext
+    {
+        public virtual DbSet<Game> Games { get; set; } = default!;
+
+        public virtual DbSet<Comment> Comments { get; set; } = default!;
+
+        public virtual DbSet<Genre> Genres { get; set; } = default!;
+
+        public virtual DbSet<PlatformType> PlatformTypes { get; set; } = default!;
+
```
Comment (ataromoku) [inline OnlineGameStore.Infrastructure/Context/GamesContext.cs:15]: Please add one more empty line after properties

File: OnlineGameStore.Infrastructure/Context/GamesContext.cs [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+        }
+
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            base.OnModelCreating(modelBuilder);
+
+            modelBuilder.ApplyConfigurationsFromAssembly(typeof(GamesContext).Assembly);
+        }
+
+        public Task CommentRepository(Comment comment)
```
Comment (ataromoku) [inline OnlineGameStore.Infrastructure/Context/GamesContext.cs:28]: What is this one for?

File: OnlineGameStore.Infrastructure/Entities/Comment.cs [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        public Guid Id { get; set; }
+
+        public string? Name { get; set; }
+
+        public string? Body { get; set; }
+
+        //public DateTime? DatePosted { get; set; }
+
+        public Guid? ParentCommentId { get; set; }
```
Comment (ataromoku) [inline OnlineGameStore.Infrastructure/Entities/Comment.cs:13]: Format the class, sort it accordingly

File: OnlineGameStore.Infrastructure/Repositories/Implementations/PlatformRepository.cs [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OnlineGameStore.Infrastructure.Repositories.Implementations
+{
+    public class PlatformRepository : RepositoryBase<PlatformType>, IPlatformRepository
+    {
+
```
Comment (ataromoku) [inline OnlineGameStore.Infrastructure/Repositories/Implementations/PlatformRepository.cs:15]: Remove empty line

File: OnlineGameStore.Infrastructure/Repositories/Implementations/RepositoryBase.cs [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+        public virtual async Task<TEntity?> GetByIdAsync(Guid id)
+        {
+            return await _forumContext.Set<TEntity>()
+                .FindAsync(id);
+        }
+
+        public virtual async Task<TEntity> AddAsync(TEntity entity)
+        {
+            _forumContext.Set<TEntity>().Add(entity);
+            await _forumContext.SaveChangesAsync();
```
Comment (ataromoku) [inline OnlineGameStore.Infrastructure/Repositories/Implementations/RepositoryBase.cs:34]: So you using save method in unit of work, but doublecalling it here?

File: OnlineGameStore.Infrastructure/Repositories/Implementations/RepositoryBase.cs [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            var record = await _forumContext.Set<TEntity>().FindAsync(id);
+
+            if (record is null)
+            {
+                return false;
+            }
+
+            _forumContext.Set<TEntity>().Remove(record);
+            await _forumContext.SaveChangesAsync();
```
Comment (ataromoku) [inline OnlineGameStore.Infrastructure/Repositories/Implementations/RepositoryBase.cs:49]: And here?

**Comment by ataromoku** [comment]
Can't see migrations

File: OnlineGameStore.Infrastructure/Repositories/Implementations/RepositoryBase.cs (hunk 23 above)
Comment (yu-krugliak) [reply OnlineGameStore.Infrastructure/Repositories/Implementations/RepositoryBase.cs]: Fixed

File: OnlineGameStore.Infrastructure/Repositories/Implementations/RepositoryBase.cs (hunk 24 above)
Comment (yu-krugliak) [reply OnlineGameStore.Infrastructure/Repositories/Implementations/RepositoryBase.cs]: Fixed

File: OnlineGameStore.Api/Controllers/GameController.cs (hunk 2 above)
Comment (yu-krugliak) [reply OnlineGameStore.Api/Controllers/GameController.cs:27]: Added

File: OnlineGameStore.Api/Program.cs (hunk 3 above)
Comment (yu-krugliak) [reply OnlineGameStore.Api/Program.cs]: Added

File: OnlineGameStore.Application/Services/Implementation/GameService.cs (hunk 8 above)
Comment (yu-krugliak) [reply OnlineGameStore.Application/Services/Implementation/GameService.cs]: Renamed

File: OnlineGameStore.Application/Services/Implementation/GameService.cs (hunk 10 above)
Comment (yu-krugliak) [reply OnlineGameStore.Application/Services/Implementation/GameService.cs]: Edited

File: OnlineGameStore.Infrastructure/Context/GamesContext.cs (hunk 20 above)
Comment (yu-krugliak) [reply OnlineGameStore.Infrastructure/Context/GamesContext.cs]: Removed

**Comment by yu-krugliak** [comment]
> Can't see migrations

Migrations added

**Comment by yu-krugliak** [comment]
Please see my last commit and comment on the unit tests, how to make them better. I know that I'm mocking the mapper incorrectly in test methods and would like to ask for advice. After your comments, I will finish all the unit tests. Thanks:)

File: OnlineGameStore.Application.Tests/Service/CommentServiceTests.cs [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+            listCommentsViews.Append(
+                new CommentView()
+                {
+                    Id = Guid.NewGuid(),
+                    Name = "Scott",
+                    GameId = game.Id
+                }
+            );
+
+            _commentRepository.Setup(x => x.GetAllAsync()).Returns(Task.FromResult(listComments));
```
Comment (ataromoku) [inline OnlineGameStore.Application.Tests/Service/CommentServiceTests.cs:76]: ReturnsAsync()?

File: OnlineGameStore.Application/Services/Implementation/GameService.cs [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+            var game = await GetExistingEntityById(gameId);
+
+            var gameView = _mapper.Map<GameView>(game);
+            return gameView;
+        }
+
+        public async Task<GameView> AddAsync(GameRequest gameRequest)
+        {
+            var game = _mapper.Map<Game>(gameRequest);
+            game.Key = Guid.NewGuid();
```
Comment (ataromoku) [inline OnlineGameStore.Application/Services/Implementation/GameService.cs:63]: We discussed this earlier

File: OnlineGameStore.Infrastructure/Startup.cs [hunk 27]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace OnlineGameStore.Infrastructure
+{
+    public static class Startup
+    {
+        public static IServiceCollection AddPersistence(this IServiceCollection services, IConfiguration configuration)
+        {
+            return services.AddDbContext<GamesContext>((provider, options) =>
+            {
+                var connectionString = configuration.GetConnectionString("DbConnection");
```
Comment (ataromoku) [inline OnlineGameStore.Infrastructure/Startup.cs:16]: Move config key to constant, bind config to object

**Review by ataromoku — CHANGES_REQUESTED** [review]

File: OnlineGameStore.Application.Tests/Service/CommentServiceTests.cs [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            _mapperMock = new Mock<IMapper>();
+            _commentRepository = new Mock<ICommentRepository>();
+            _sut = new CommentService(_commentRepository.Object, _mapperMock.Object);
+        }
+
+        [Fact]
+        public async Task GetCommentsByGameId_WhenListIsEmpty_ThenEmptyListReturned()
+        {
+            var game = new Game() { Id = Guid.NewGuid() };
```
[the remark this answers is not in the document]
Comment (yu-krugliak) [reply OnlineGameStore.Application.Tests/Service/CommentServiceTests.cs:29]: Will change this in next PR

File: OnlineGameStore.Application.Tests/Service/CommentServiceTests.cs (hunk 25 above)
Comment (yu-krugliak) [reply OnlineGameStore.Application.Tests/Service/CommentServiceTests.cs:76]: Thanks, will change this in next PR

File: OnlineGameStore.Application.Tests/Service/CommentServiceTests.cs [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+            _commentRepository.Verify(x => x.GetByGameId(game.Id), Times.Once);
+            _commentRepository.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task GetCommentsByGameId_WhenListIsNotEmpty_ThenListReturned()
+        {
+            var game = new Game() { Id = Guid.NewGuid() };
+
+            IEnumerable<Comment> listComments = Enumerable.Empty<Comment>();
```
[the remark this answers is not in the document]
Comment (yu-krugliak) [reply OnlineGameStore.Application.Tests/Service/CommentServiceTests.cs:42]: Will change this in next PR

File: OnlineGameStore.Api/Program.cs [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+builder.Services.AddEndpointsApiExplorer();
+builder.Services.AddSwaggerGen();
+
+builder.Services.AddCors(p => p.AddPolicy("corsapp", policyBuilder =>
+{
+    policyBuilder.AllowAnyOrigin().AllowAnyMethod().AllowAnyHeader();
+}));
+
+builder.Services.AddPersistence(builder.Configuration);
```
[the remark this answers is not in the document]
Comment (yu-krugliak) [reply OnlineGameStore.Api/Program.cs:20]: Done

File: OnlineGameStore.Infrastructure/Startup.cs (hunk 27 above)
Comment (yu-krugliak) [reply OnlineGameStore.Infrastructure/Startup.cs]: Done, thnx

File: OnlineGameStore.Infrastructure/Entities/Genre.cs [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+        public Guid Id { get; set; }
+
+        public string? Name { get; set; }
+
+        public string? Description { get; set; }
+
+
+        public Guid? ParentGenreId { get; set; }
+        public virtual Genre? ParentGenre { get; set; }
+
```
Comment (ataromoku) [inline OnlineGameStore.Infrastructure/Entities/Genre.cs:14]: Format properly pllzzz

**Review by ataromoku — CHANGES_REQUESTED** [review]

File: OnlineGameStore.Application/Models/Requests/GameRequest.cs [hunk 32]
```diff
@@ -0,0 +1,14 @@
+﻿namespace OnlineGameStore.Application.Models.Requests
+{
+    public class GameRequest
+    {
+        public string? Name { get; set; }
+
+        public string? Description { get; set; }
+
```
Comment (ataromoku) [inline OnlineGameStore.Application/Models/Requests/GameRequest.cs:8]: I think you accidentally removed the `key` property :\

**Review by ataromoku — CHANGES_REQUESTED** [review]

File: OnlineGameStore.Application/Models/Requests/GameRequest.cs (hunk 32 above)
Comment (yu-krugliak) [reply OnlineGameStore.Application/Models/Requests/GameRequest.cs:10]: Yes, thnx, a added key back

Comment (ataromoku) [reply OnlineGameStore.Application/Models/Requests/GameRequest.cs:10]: Key must be string

**Review by ataromoku — CHANGES_REQUESTED** [review]

**Comment by ataromoku** [comment]
Add migrations and feel free to merge

**Review by ataromoku — APPROVED** [review]

**Pull request merged**